disktable.go

```diff
@@ -0,0 +1,650 @@
+		matches := btree.New(2)
+
+		// fetch all rows for each index that match our lookup constraint.
+		// Record those rows into a 2-3-4 tree. Any entry in the tree that has
+		// len(lookup) as its value is a line we want to return.
+		errCh := make(chan error, 1)
+		for _, l := range lookup {
+			l := l
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for result := range t.indexLookup(l.IndexName, l.Key) {
+					if ctx.Err() != nil {
+						return
+					}
+					if result.err != nil {
+						if result.err == badger.ErrKeyNotFound {
+							continue
+						}
+						cancel()
+						select {
+						case errCh <- result.err:
+						default:
+						}
+					}
+					// Add entry to btree or increase existing entry counter.
+					mu.Lock()
+
+					var counter btreeItem
+					a := matches.Get(btreeItem{primaryKey: result.i})
+					if a == nil {
+						counter = btreeItem{primaryKey: result.i}
+					} else {
+						counter = a.(btreeItem)
+					}
+					counter.count++
+					matches.ReplaceOrInsert(counter)
+
+					mu.Unlock()
+				}
+			}()
+		}
+
+		wg.Wait()
+
+		// Did we have any errors, if so return them.
+		select {
+		case err := <-errCh:
+			ch <- Result{Err: err}
+			return
+		default:
+		}
+
+		keysLookup := make([][]byte, 0, 100)
+
+		// Pull all the rows in the index that matched and send them back.
+		matches.Ascend(
+			func(item btree.Item) bool {
+				i := item.(btreeItem)
+				if int(i.count) == len(lookup) {
+					primaryKey := make([]byte, 8)
+					endian.PutUint64(primaryKey, i.primaryKey)
+					keysLookup = append(keysLookup, primaryKey)
+					// Ok, we have enough for a bulk lookup.
+					if len(keysLookup) == 1000 {
+						t.primary.db.View(t.retrievePrimaryRows(keysLookup, ch))
+						keysLookup = keysLookup[0:0]
+					}
+				}
+				return true
+			},
+		)
+
+		// Get any outstanding lookups that didn't have enough to do a bulk.
+		if len(keysLookup) > 0 {
+			t.primary.db.View(t.retrievePrimaryRows(keysLookup, ch))
+		}
+	}()
+
+	return ch, nil
+}
+
+func (t *Table) retrievePrimaryRows(keys [][]byte, ch chan Result) func(txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		for _, key := range keys {
+			item, err := txn.Get(key)
+			if err != nil {
+				return err
+			}
+			if !bytes.Equal(item.Key(), key) {
+				panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+			}
+			value, err := item.ValueCopy(nil)
+			if err != nil {
+				return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+			}
+			ch <- Result{Value: value}
+		}
+		return nil
+	}
+}
+
+type lookupResult struct {
+	i   uint64
+	err error
+}
+
+// indexLookup returns all matching keys in an index.
+func (t *Table) indexLookup(indexName string, key []byte) chan lookupResult {
+	index := t.indexByName[indexName]
+
+	realKey := finalKey(key, 0)
+
+	ch := make(chan lookupResult, 1)
+	if index.AllowDuplicates {
+		// multiMatch will close the channel.
+		go t.multiMatch(realKey, index, ch)
```

Comment:
long term, I think it's gonna be better to close the channel here, as close to where it's created as possible